File: 2107-find-unique-binary-string/find-unique-binary-string.cpp
class Solution {
public:
    string findDifferentBinaryString(vector<string>& nums) {
        unordered_map<string,int>map;
        int n=nums.size();
        string str="";
        for(int i=0;i<n;i++){
            map[nums[i]]++;
            str.push_back('0');
        }
        cout<<str<<endl;
        string ans;
        fun(0,str,ans,map);
        return ans;
    }

    bool fun(int ind,string str,string &ans,unordered_map<string,int>&map){
        if(map.find(str)==map.end()){
            ans=str;
            return true;
        }
        if(ind==str.size()){
            return false;
        }
        str[ind]='1';
        if(fun(ind+1,str,ans,map)){
            return true;
        }
        str[ind]='0';
        return false;
    }
};

File: 2107-find-unique-binary-string/find-unique-binary-string.java
class Solution {
    public String findDifferentBinaryString(String[] nums) {
        StringBuilder sb = new StringBuilder();
        Map<String,Integer> map = new HashMap<>();
        int n = nums.length;
        for(int i=0;i<n;i++){
            map.put(nums[i],i);
        }
        return fun(0,sb,map,n);
    }

    String fun(int ind,StringBuilder subAns, Map<String,Integer>map,int n){

        if(ind==n){
            if(!map.containsKey(subAns.toString())){
                return subAns.toString();
            }
            return "";
        }

        String str1 = fun(ind+1,subAns.append('0'),map,n);
        if(!str1.equals("")){
            return str1;
        }
        subAns.deleteCharAt(subAns.length()-1);
        String str2 = fun(ind+1,subAns.append('1'),map,n);
        if(!str2.equals("")){
            return str2;
        }
        subAns.deleteCharAt(subAns.length()-1);
        return "";
    }
}